lightning/src/blinded_path/payment.rs

```diff
@@ -35,6 +34,74 @@ use core::ops::Deref;
 #[allow(unused_imports)]
 use crate::prelude::*;
 
+/// A blinded path to be used for sending or receiving a payment, hiding the identity of the
+/// recipient.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct BlindedPaymentPath(pub(crate) BlindedPath);
+
+impl Writeable for BlindedPaymentPath {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		self.0.write(w)
+	}
+}
+
+impl Readable for BlindedPaymentPath {
+	fn read<R: io::Read>(r: &mut R) -> Result<Self, DecodeError> {
+		Ok(Self(BlindedPath::read(r)?))
+	}
+}
+
+impl BlindedPaymentPath {
+	/// Create a blinded path for a payment, to be forwarded along `intermediate_nodes`.
+	///
+	/// Errors if:
+	/// * a provided node id is invalid
+	/// * [`BlindedPayInfo`] calculation results in an integer overflow
+	/// * any unknown features are required in the provided [`ForwardTlvs`]
+	//  TODO: make all payloads the same size with padding + add dummy hops
+	pub fn new<ES: Deref, T: secp256k1::Signing + secp256k1::Verification>(
+		intermediate_nodes: &[ForwardNode], payee_node_id: PublicKey, payee_tlvs: ReceiveTlvs,
+		htlc_maximum_msat: u64, min_final_cltv_expiry_delta: u16, entropy_source: ES,
+		secp_ctx: &Secp256k1<T>
+	) -> Result<(BlindedPayInfo, Self), ()> where ES::Target: EntropySource {
```

Comment:
For some reason I assumed we would make `BlindedPayInfo` a part of `BlindedPaymentPath` with this refactor. Did we decide against that?

Reply:
Hmm, that wasn't my impression. Maybe @TheBlueMatt wants to weigh in.

Reply:
Started looking into this. Preliminary thoughts are that it looks feasible if we add a `BlindedPaymentPath::writeable_blinded_path` method that returns a `Writeable` wrapper around `BlindedPath` (so that the inner blinded path type can remain private). 

Getting quite a few compiler errors so would prefer to save this for follow-up if we decide to go this way. @jkczyz do you prefer this approach?

Reply:
> Started looking into this. Preliminary thoughts are that it looks feasible if we add a `BlindedPaymentPath::writeable_blinded_path` method that returns a `Writeable` wrapper around `BlindedPath` (so that the inner blinded path type can remain private).

Would making the inner fields `pub(crate)` be sufficient?

> Getting quite a few compiler errors so would prefer to save this for follow-up if we decide to go this way. @jkczyz do you prefer this approach?

Sure, follow-up is fine by me.

> I guess I still don't get why "having it implement Writeable might cause us to inadvertently write the wrong data." I don't think this PR runs that risk?

I was referring to the case where `BlindedPaymentPath` includes both `BlindedPath` and `BlindedPayinfo`. There we wouldn't want to make `BlindedPaymentPath` implement `Writeable` as we could write the wrong data for the `paths` TLV in `Bolt12Invoice`.

Reply:
There are only test vectors for `offer` messages lol

Reply:
>> Started looking into this. Preliminary thoughts are that it looks feasible if we add a BlindedPaymentPath::writeable_blinded_path method that returns a Writeable wrapper around BlindedPath (so that the inner blinded path type can remain private).

> Would making the inner fields pub(crate) be sufficient?

Not sure I'm following. Do you mean the inner fields of `BlindedPath`? The goal is to make that struct `pub(super)`. 

Here's how the public path will look from my PoV:
```
pub struct BlindedPaymentPath {
    inner: BlindedPath,
    pub payinfo: BlindedPayInfo
}

Reply:
> Not sure I'm following. Do you mean the inner fields of `BlindedPath`? The goal is to make that struct `pub(super)`.
> 
> Here's how the public path will look from my PoV:
> 
> ```
> pub struct BlindedPaymentPath {
>     inner: BlindedPath,
>     pub payinfo: BlindedPayInfo
> }
> ```

I'm not too opinionated on how it's done, but was just pointing out the internals need to be accessible in `offers::invoice` to serialized as two separate TLVs.